test/func_ds3_lock_readlckd.py

```diff
@@ -0,0 +1,249 @@
+    printf("FAIL: Missing argument (primary|secondary)\n");
+    return -2;
+  }
+  if (strcmp(argv[1], "primary") == 0)
+    primary = 1;
+  if (strcmp(argv[1], "secondary") == 0)
+    primary = 0;
+  if (primary < 0) {
+    printf("FAIL: Invalid argument (primary|secondary)\n");
+    return -2;
+  }
+
+  if (primary) {
+    unsigned int attr;
+
+    if (_dos_creatnew(FNAME, 0, &handle) != 0) {
+      printf("FAIL: %s: File '%s' not created\n", argv[1], FNAME);
+      return -1;
+    }
+    if (_dos_write(handle, FDATA, strlen(FDATA), &rc) != 0) {
+      printf("FAIL: %s: File '%s' not written to\n", argv[1], FNAME);
+      _dos_close(handle);
+      return -1;
+    }
+    if (_dos_close(handle) != 0) {
+      printf("FAIL: %s: File '%s' not closed\n", argv[1], FNAME);
+      return -1;
+    }
+
+    if (_dos_getfileattr(FNAME, &attr)) {
+      printf("FAIL: %s: File '%s' getfileattr()\n", argv[1], FNAME);
+      return -1;
+    }
+    if (attr != _A_ARCH) {
+      printf("FAIL: %s: File '%s' attrs incorrect\n", argv[1], FNAME);
+      printf("      Attrs are: (0x%02x)\n", attr);
+      if (attr & _A_ARCH)
+        printf("                Archive\n");
+      if (attr & _A_RDONLY)
+        printf("                Read only\n");
+      if (attr & _A_HIDDEN)
+        printf("                Hidden\n");
+      if (attr & _A_SYSTEM)
+        printf("                System\n");
+      if (attr & _A_VOLID)
+        printf("                Volume ID\n");
+      if (attr & _A_SUBDIR)
+        printf("                Directory\n");
+      return -1;
+    }
+
+    ret = _dos_open(FNAME, O_RDONLY, &handle);
+    if (ret != 0) {
+      printf("FAIL: %s: File '%s' not opened(%d)\n", argv[1], FNAME, ret);
+      return -1;
+    }
+
+    if (_dos_lock(handle, 5,  3) != 0) { // somewhere in the middle
+      printf("FAIL: %s: Could not get lock on file '%s'\n", argv[1], FNAME);
+      _dos_close(handle);
+      return -1;
+    }
+
+    printf("OKAY: %s: Acquired lock on file '%s'\n", argv[1], FNAME);
+
+    // Now start second copy
+    spawnlp(P_WAIT, argv[0], argv[0], "secondary", NULL);
+
+    _dos_unlock(handle, 5,  3);
+    _dos_close(handle);
+
+  } else { // secondary
+    char buf[80];
+
+    ret = _dos_open(FNAME, O_RDONLY, &handle);
+    if (ret != 0) {
+      printf("FAIL: %s: File '%s' not opened(%d)\n", argv[1], FNAME, ret);
+      return -1;
+    }
+
+    // Check what was written by primary process
+
+    // partially overlapping range prior to lock (should fail)
+    memset(buf, ' ', strlen(FDATA));
+    ret = _dos_read(handle, buf, 6, &rc);
+    if (ret == 0) {
+      printf("WARN: %s: File '%s' overlap conflict(1) read by mistake cnt = %d\n", argv[1], FNAME, rc);
+      // Stas wants to allow this 'enhancement', so reset the file position
+      if (llseek(handle, 0, SEEK_SET) != 0) {
+        printf("FAIL: %s: File '%s' seek failed\n", argv[1], FNAME);
+        _dos_close(handle);
+        return -1;
+      }
+    } else if (ret != 5) {
```

Comment:
`ret` or `rc`?
A few lines below you compare rc to 5...

Reply:
Looks dubious, will check.

Reply:
Is correct, I thought we were in a different file. ret is dos return code (5 == ACCESS_DENIED), rc is number of chars read.

Reply:
And I forgot to say, the rc is being checked for the number of chars read, which unfortunately due to my choice of test data, is also 5!

Reply:
OK but you still don't check rc==5
when ret==0, i.e. that the "extension"
worked correctly and only allowed 5 bytes,
but not 6.

Reply:
Indeed you are correct although that section is information only (no pass or fail generated from the extension now), so the most I can do as is, is print it to screen. However unless the user is running the test by hand it will never appear anywhere unless the test fails for some later reason. I guess my heart wasn't really in testing the extension properly as it was a departure from MS-DOS 6.22.  Do you want me to add something in a follow up commit?

Reply:
Yes, would be nice to fix this.
Otherwise you miss the completely
broken case of just reading the locked
region! "extension" is not supposed to
break things, locked region should never
be read.

Reply:
okay, I'll work on that tomorrow.

Reply:
Any progress?

Reply:
I'm a little busy with other things at the moment. I rebased it, but I can't quite remember whether it was actually finished or not. https://github.com/dosemu2/dosemu2/pull/1339

Reply:
OK, applied.